drivers/sensor/si1153/si1153.c

```diff
@@ -0,0 +1,663 @@
+
+	/* Right channel LED3 */
+	retval += si115x_param_set(data, PARAM_ADCCONFIG3 , ADCCONFIG);
+	retval += si115x_param_set(data, PARAM_ADCSENS3   , ADCSENS);
+	retval += si115x_param_set(data, PARAM_ADCPOST3   , ADCPOST);
+	retval += si115x_param_set(data, PARAM_MEASCONFIG3, MEASURECFG | MEASCFG_LED3_ENA);
+
+	retval += si115x_param_set(data, PARAM_LED1_A, LED_CURRENT_199);
+	retval += si115x_param_set(data, PARAM_LED2_A, LED_CURRENT_199);
+	retval += si115x_param_set(data, PARAM_LED3_A, LED_CURRENT_310);
+
+	retval += si1153_reg_write(data, SI115x_REG_IRQ_ENABLE, 0x0f);
+
+	return retval;
+}
+
+/**
+ *
+ */
+static void si115x_handler(struct si1153_data *data, struct SI115X_SAMPLES *samples)
+{
+	uint8_t buffer[13];
+	int irq_status;
+	int count;
+
+	/* wait for interrupt signaling the registers are update */
+	count = 0;
+	do {
+		irq_status = si1153_reg_read(data, SI115x_REG_IRQ_STATUS);
+		k_busy_wait(50);
+		count++;
+	} while ((count < 20) && ((irq_status & 0xf) != 0xf));
+	if (count >= 40) {
+		printk("SI1153:ERROR interrupt timeout\n");
+		return;
+	}
+	si1153_block_read(data, SI115x_REG_IRQ_STATUS, 13, buffer);
+
+	samples->irq_status = irq_status; //buffer[0];
+
+	samples->timestamp = (u32_t)k_uptime_get();
+
+	samples->ch0  = buffer[1] << 16;
+	samples->ch0 |= buffer[2] <<  8;
+	samples->ch0 |= buffer[3];
+	if (samples->ch0 & 0x800000) {
+		samples->ch0 |= 0xFF000000;
+	}
+
+	samples->ch1  = buffer[4] << 16;
+	samples->ch1 |= buffer[5] <<  8;
+	samples->ch1 |= buffer[6];
+	if (samples->ch1 & 0x800000) {
+		samples->ch1 |= 0xFF000000;
+	}
+
+	samples->ch2  = buffer[7] << 16;
+	samples->ch2 |= buffer[8] <<  8;
+	samples->ch2 |= buffer[9];
+	if (samples->ch2 & 0x800000) {
+		samples->ch2 |= 0xFF000000;
+	}
+
+	samples->ch3  = buffer[10] << 16;
+	samples->ch3 |= buffer[11] <<  8;
+	samples->ch3 |= buffer[12];
+	if (samples->ch3 & 0x800000) {
+		samples->ch3 |= 0xFF000000;
+	}
+
+}
+
+static int si1153_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+#ifndef CONFIG_SI1153_OWN_THREAD
+	struct si1153_data *data = dev->driver_data;
+	struct SI115X_SAMPLES      samples;
+
+	si115x_force(data);
+
+	si115x_handler(data, &samples);
+
+	data->gesture = si115x_gesture_algorithm(&samples);
+	data->ch0 = samples.ch0;
+	data->ch1 = samples.ch1;
+	data->ch2 = samples.ch2;
+	data->ch3 = samples.ch3;
+
+#if 1
+	char*strgesture[] = {"NONE", "UP", "DOWN", "LEFT", "RIGHT", "PROX"};
+	printk("Gesture %d %s %d %d %d %d\n",
+			data->gesture, strgesture[data->gesture],
+			data->ch0,
+			data->ch1,
+			data->ch2,
+			data->ch3);
+#endif
+
+#endif
+	return 0;
+}
+
+static int si1153_channel_get(struct device *dev,
+					  enum sensor_channel chan,
+					  struct sensor_value *val)
+{
+	struct si1153_data *data = dev->driver_data;
+
+	switch (chan) {
+
+	case SENSOR_CHAN_LIGHT:
+	case SENSOR_CHAN_IR:
+		val->val1 = data->ch0;
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_PROX:
```

Comment:
The Sensor API doesn't take into account multiple channels, so this behavior would be against what an application developer would expect.